### PR TITLE
Enh/parametrised perms

### PR DESCRIPTION
**Pull request by kalmarek**
Parametrise permutations over subtypes of Integers to save on storage and allocations;
`G = PermutationGroup(Int16(10))` produces parent which has only `perm{Int16}` as elements;
You can coerce `p::perm{Int16}` to `perm{I}` (where `I<:Integer`), e.g. by `PermutationGroup(I(10))(p)`.

The new `cycles` code is faster (as it allocates once), but for small `N` (ie. `<15`). There is however a certain slowdown for powering (via `^`, using cycles) for sizes up to `90` which I can not explain. Anyone?

Testing (all functions are replaced by their not-caching copies): `TRIALS` medians are collected of `@elapsed f(p) ` and their distribution plotted as violin plot.
```
N = 500
TRIALS=500
SAMPLESIZE=10_000 #for inv, parity
SAMPLESIZE2=1000 #cycles, multiplication, ^

function perm_sample(N::T=500; samplesize=1000, seed=1234) where T<:Integer
    srand(seed)
    return [rand(PermutationGroup(N)) for i in 1:samplesize]
end

function bench(f; N=N samplesize=SAMPLESIZE, trials=TRIALS, seed=1234)
    all = zeros(trials)
    srand(seed)
    samples = perm_sample(N, samplesize=samplesize)
    for i in 1:trials
        all[i] = median([@elapsed f(p) for p in samples])
    end
    return all
end
```
Speedups: (`median(bench(f, Int64(N))) = 1.0`):
newplot
newplot 1
newplot 2
newplot 3
index
index

If You have other changes in mind which can be included in this pull feel free to comment.

### CODE REVIEW

**Comment by thofma** [comment]
This looks good, thanks! I have no idea about the slowdown.

I am happy to merge this once the tests pass (they are failing currently).

**Comment by kalmarek** [comment]
Ooops....  give me a few minutes ;-)

**Comment by wbhart** [comment]
Yeah, those are some really nice speedups. As far as I can see, this shows
that the actual arithmetic itself is slower with the Int16. This is
probably due to the actual processor itself (unaligned access, legacy
support for 16 bit machines, etc.) However, the cache benefits for large
permutations is clear.

Apparently Gap internally chooses automatically between representations
depending on size. But I don't know why, except that probably 16 bits is
just not enough for some purposes. It's not too trivial to test whether the
same slowdown happens there.

By the way, Gap does composition from the right, rather than the left, as
you have, i.e. a*b would apply b to the result of a. Are you certain you
want to stick with the convention you have chosen?

**Comment by wbhart** [comment]
By the way, can you do all future PR's for group theory stuff that doesn't
make use of Flint against the new AbstractAlgebra.jl [1] (as soon as I move
your latest changes over to there, of course).

AbstractAlgebra.jl is the same as Nemo, except it only has generic Julia
code and nothing from Flint, Arb or Antic. Nemo will use AbstractAlgebra.jl
in future, rather than have this code directly in it.

Of course if you implement stuff that does need Flint, Arb or Antic, by all
means, please commit this against Nemo (it's not immediately clear where to
put it until we remove the existing generic stuff from Nemo, but we'll do
that in a relatively short amount I time I think).

[1] https://github.com/Nemocas/AbstractAlgebra.jl/

**Comment by kalmarek** [comment]
@wbhart Your explanation of `Int16` vs `Int64` makes sense, thanks (although my knowledge on caches, etc is very limited)

It is somehow natural (for me) to think of permutations as functions; then `a*b` at `i` becomes `a(b(i))`. However, I don't perform any particular calculations with particular group elements so I am open to change this convention. What do others think about composition of functions vs compatibility with GAP?

Is actually choosing the type automatically based on value type-stable? My impression is that the compiler wouldn't be able to infer the type of `PermutationGroup(N)` at compile time? 
GAP [switches between Int16 and Int32](https://www.gap-system.org/Manuals/doc/ref/chap42.html#X80F07BE2811D4BAC) based on the maximal index of point moved by a particular permutation.

File: src/generic/GenericTypes.jl [hunk 1]
```diff
@@ -12,14 +12,14 @@
...
 
-mutable struct PermGroup <: Nemo.Group
-   n::Int
+mutable struct PermGroup{T<:Integer} <: Nemo.Group
+   n::T
 
-   function PermGroup(n::Int, cached=true)
+   function PermGroup(n::T, cached=true) where T<:Integer
       if haskey(PermID, n)
          return PermID[n]::PermGroup
```
Comment (thofma) [inline src/generic/GenericTypes.jl]: If you change it to `::PermGroup{T}`, it will be type stable.

**Comment by thofma** [comment]
I think it is fine. `PermGroup(x::T)` will give something of type `PermGroup{T}`.

**Comment by wbhart** [comment]
What do @fieker and @thofma think about the order of composition?

**Comment by wbhart** [comment]
Regarding switching types based on value, I wasn't suggesting we do that. Though I see no problem with it, given that only users at the top level (where things aren't compiled anyway) are supposed to use those constructors in the first place.

On the other hand, permutations are so fundamental that it is pretty likely that it's going to be hard to enforce that people use the internal constructors only (where a type must be specified), when writing library code. That's going to be the rule, but probably it will be hard to enforce in this case, which might actually introduce some type instability.

**Comment by kalmarek** [comment]
@thofma Currently `PermutationGroup(n)` will give You `PermGroup{typeof(n)}`, based on the type of `n` just fine. I understood that we're considering something like (to shield user from picking the 'economical` type)
```julia
function PermGroup(n)
   for T in [Int8, Int16, Int32, Int64]
      if n<typemax(T)
         return Generic.PermGroup(T(n))
      end
   end
   return (...default...)
end
```
This `PermGroup(n)` wouldn't be type-stable?

**Comment by wbhart** [comment]
This function as you have it here will return something based on the value
of n, not the type of n, so it's not type stable. I definitely wouldn't do
this without a really compelling reason.

**Comment by thofma** [comment]
Sorry, I misread something. Now I understand and I don't think we should do it. It should be `typeof(PermutationGroup(n)) = PermGroup{typeof(n)}`.

But also with the current version, there is a type instablity as I pointed out above. Can you fix this?

Regarding the composition, @fieker and I think it is best to just go with what you did here. We can change it later if necessary.

**Comment by thofma** [comment]
Thanks. I'll let travis/appveyor run more time and then I'll merge this.

File: src/generic/GenericTypes.jl (hunk 1 above)
Comment (kalmarek) [reply src/generic/GenericTypes.jl]: Did You mean typeassert `PermID[n]::PermGroup{T}`?

Comment (thofma) [reply src/generic/GenericTypes.jl]: Yes.

**Comment by wbhart** [comment]
I've now manually merged these changes into AbstractAlgebra.jl [1]. In future, could you make all PR's directly against AbstractAlgebra.jl (which will be required by Nemo, so everything will still be available from Nemo).

Note: Nemo.Group => AbstractAlgebra.Group 

https://github.com/Nemocas/AbstractAlgebra.jl/

**Comment by kalmarek** [comment]
Thanks;

I have code for parsing GAP permutations lying somewhere around. Should I push it there?
(I won't switch my code to splitted `AbstractAlgebra` right now)

**Comment by wbhart** [comment]
Yes, you should make the PR for that against AbstractAlgebra.jl not
Nemo.jl, assuming you don't need to do it in the next few days. We won't
accept PR's for Generic stuff against Nemo.jl any more.

(Until the split is complete, we actually need it in both Nemo.jl and
AbstractAlgebra.jl. This will only be the case for a couple of days to
weeks at most though, I hope. So perhaps you can wait until the split is
done, or if you want it in Nemo as soon as possible, make the PR against
both Nemo.jl and AbstractAlgebra.jl. For your code the changes will be
identical, except Nemo.Group becomes AbstractAlgebra.Group in
AbstractAlgebra.jl. All your group theory code is already working in
AbstractAlgebra.jl, of course.)

**Pull request merged**